dango/lending/src/query.rs

```diff
@@ -64,3 +77,62 @@ pub fn query_debts(
         .take(limit as usize)
         .collect()
 }
+
+pub fn query_collateral_powers(
+    storage: &dyn Storage,
+) -> StdResult<BTreeMap<Denom, CollateralPower>> {
+    COLLATERAL_POWERS.load(storage)
+}
+
+/// Calculates the health of a margin account.
+pub fn calculate_account_health(
+    querier: &QuerierWrapper,
+    margin_account: Addr,
+    debts: Coins,
+    collateral_powers: BTreeMap<Denom, CollateralPower>,
+) -> anyhow::Result<HealthResponse> {
+    // Calculate the total value of the debts.
+    let mut total_debt_value = Udec128::ZERO;
+    for coin in debts {
+        let price = dango_oracle::raw_query_price(querier, &coin.denom)?;
+        total_debt_value += price.value_of_unit_amount(coin.amount);
+    }
+
+    // Calculate the total value of the account's collateral adjusted for the collateral power.
+    let mut total_adjusted_collateral_value = Udec128::ZERO;
+    for (denom, power) in collateral_powers {
+        let collateral_balance = querier.query_balance(margin_account, denom.clone())?;
+
+        // As an optimization, don't query the price if the collateral balance is zero.
+        if collateral_balance.is_zero() {
+            continue;
+        }
+
+        let price = dango_oracle::raw_query_price(querier, &denom)?;
+        let collateral_value = price.value_of_unit_amount(collateral_balance);
+        total_adjusted_collateral_value += collateral_value * power.into_inner();
+    }
```

Comment:
if there's an asset that's both a collateral and a debt, then we'd be repeat the oracle query twice

Reply:
also every time you do `dango_oracle::raw_query_price`, you do `query_app_config` again, which is a waste...

Reply:
> if there's an asset that's both a collateral and a debt, then we'd be repeat the oracle query twice

Yeah that's what I meant by creating a querier that caches results. Thought you said it was not necessary. Can implement that if you want?